### PR TITLE
#36 Validate ips with ports

### LINKED ISSUE #36: NodeCommunicationClient can't parse IPs, because ports can't go through the Validator
**Describe the bug**
In order for NodeCommunicationClient to parse the ip from ConfigurationTool it needs to have an ip. If the passed address (from args) is localhost, the Validator (Uri.CheckHostName specifically) counts it as dns, but if it's 127.0.0.1 as IPv4. If we now pass a port (localhost:1234 or 127.0.0.1:1234), both become Unknown and an InvalidIpOrHostnameException is thrown.
If we don't pass the port, NodeCommunicationClient tries to parse it if there were a port (splits it by ':' and the port is the second element in the array) and that throws IndexOutOfRangeException

**To Reproduce**
You don't even have to start the master node  
1. dotnet run --project Olymp/Olymp.csproj --conf localhost -u admin -p admin

**Expected behavior**
The Configuration tool would start listening on the adress, instead of throwing

**Screenshots**
Screenshot from 2019-07-16 11-59-29
Screenshot from 2019-07-16 11-59-16

**Desktop (please complete the following information):**
 - OS: Linux x64
 - Dotnet version: 2.2.300

**Additional context**
I'm too lazy to file another bug report, but that shouldn't be happening:
Screenshot from 2019-07-16 11-58-56

**Possible fixes**
We can either:
1. Allow ports in Validator.ValidateAddress
2. Set default port in ConfigurationTool constructor
3. Set default port in NodeCommunicationClient

### PR DESCRIPTION
**Pull request by Mitiko**
# Description
In order for NodeCommunicationClient to parse the ip from ConfigurationTool it needs to have an ip. If the passed address (from args) is localhost, the Validator (Uri.CheckHostName specifically) counts it as dns, but if it's 127.0.0.1 as IPv4. If we now pass a port (localhost:1234 or 127.0.0.1:1234), both become Unknown and an InvalidIpOrHostnameException is thrown.
If we don't pass the port, NodeCommunicationClient tries to parse it if there were a port (splits it by ':' and the port is the second element in the array) and that throws IndexOutOfRangeException

Fixes #36

## Type of change
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
With the test from the bug report to reproduce:
dotnet run --project Olymp/Olymp.csproj --configure localhost --user admin --password admin

**Test Configuration**:
* Hardware: Linux x64 Ubuntu
* dotnet version: 2.2.300

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: Olymp.Communication/NodeCommunicationClient.cs [hunk 1]
```diff
@@ -17,8 +18,9 @@ public static class NodeCommunicationClient
             {
                 Validator.ValidateStringValues(username, password, message.ToString());
                 var data = EncryptMessage(username, password, command, message);
-                // TODO: Parse ip adress better
-                client = new TcpClient(server.Split(":")[0], int.Parse(server.Split(":")[1]));
+
+                (string ip, int port) = Validator.ParseIpAddress(serverAddress);
```
Comment (Azer0s) [inline Olymp.Communication/NodeCommunicationClient.cs]: This should probably be a struct instead of a named tuple.

File: Olymp.Util/Validator.cs [hunk 2]
```diff
@@ -60,5 +60,23 @@ public static bool ValidateStringValues(params string[] values)
                     throw new InvalidArgumentException(value);
             return true;
         }
+
+        public static (string ip, int port) ParseIpAddress(string address)
+        {
+            var ipDeconstruction = address.Split(":");
+            string ip = ipDeconstruction[0], port;
```
Comment (Azer0s) [inline Olymp.Util/Validator.cs]: This is kinda unreadable. Please do 
```cs
var ip = ipDeconstruction[0];
var port = string.Empty;
```

File: Olymp.Communication/NodeCommunicationClient.cs (hunk 1 above)
Comment (Mitiko) [reply Olymp.Communication/NodeCommunicationClient.cs]: Do you think therre will be that much use?

File: Olymp.Util/Validator.cs (hunk 2 above)
Comment (Mitiko) [reply Olymp.Util/Validator.cs]: aaaaaaaaaaaaaaaaaaaaaaaaaaaaahhhhhhhhhhhhhhhhhhhhhh

File: Olymp.Communication/NodeCommunicationClient.cs (hunk 1 above)
Comment (Azer0s) [reply Olymp.Communication/NodeCommunicationClient.cs]: @Mitiko Depends on how often we use these values. A named tuple is fine for like 2-3 usages. Anything above that should really be a struct.

File: Olymp.Util/Validator.cs (hunk 2 above)
Comment (Azer0s) [reply Olymp.Util/Validator.cs]: @Mitiko I know...I'm starting a philosophical debate here..

File: Olymp.Communication/NodeCommunicationClient.cs (hunk 1 above)
Comment (Mitiko) [reply Olymp.Communication/NodeCommunicationClient.cs]: How about we take that into consideration  and be lazy until we have to use it some more

File: Olymp.Util/Validator.cs (hunk 2 above)
Comment (Mitiko) [reply Olymp.Util/Validator.cs]: You still wrote that code, couldn't you put it as a suggestion so I can just approve it a commit from the browser directly

File: Olymp.Communication/NodeCommunicationClient.cs (hunk 1 above)
Comment (Azer0s) [reply Olymp.Communication/NodeCommunicationClient.cs]: @Mitiko Sure thing. 😆

File: Olymp.Util/Validator.cs (hunk 2 above)
Comment (Azer0s) [reply Olymp.Util/Validator.cs]: I can do that?? 😱

Comment (Mitiko) [reply Olymp.Util/Validator.cs]: I just found out when I reviewed #37

File: Olymp.Util/Validator.cs [hunk 3]
```diff
@@ -64,7 +64,8 @@ public static bool ValidateStringValues(params string[] values)
         public static (string ip, int port) ParseIpAddress(string address)
         {
             var ipDeconstruction = address.Split(":");
-            string ip = ipDeconstruction[0], port;
+            string ip = ipDeconstruction[0];
+            string port = "";
```
Comment (Azer0s) [inline Olymp.Util/Validator.cs]: This is evil. This is the reason why climate change exists and why Trump is US president. I hate `""` and I'm not ashamed. `string.Empty` masterrace.

Comment (Mitiko) [reply Olymp.Util/Validator.cs]: You stubborn perfectionist

File: Olymp.Util/Validator.cs [hunk 4]
```diff
@@ -65,7 +65,7 @@ public static (string ip, int port) ParseIpAddress(string address)
         {
             var ipDeconstruction = address.Split(":");
             string ip = ipDeconstruction[0];
-            string port = "";
+            string port = string.Empty;
```
Comment (Azer0s) [inline Olymp.Util/Validator.cs]: ❤️

File: Olymp.Util/Validator.cs (hunk 3 above)
Comment (Azer0s) [reply Olymp.Util/Validator.cs]: @Mitiko 😘

File: Olymp.Util/Validator.cs (hunk 2 above)
Comment (Mitiko) [reply Olymp.Util/Validator.cs]: https://help.github.com/en/articles/reviewing-proposed-changes-in-a-pull-request#about-reviewing-pull-requests

**Pull request merged**